File: oathkeeper/authorizer/authorizer.py
# A simple request-based authorizer example to demonstrate how to use request
# parameters to allow or deny a request. In this example, a request is
# authorized if the client-supplied headerauth1 header, QueryString1
# query parameter, and stage variable of StageVar1 all match
# specified values of 'headerValue1', 'queryValue1', and 'stageValue1',
# respectively.

import json


def lambda_handler(event, context):
    print(event)

    # Retrieve request parameters from the Lambda function input:
    headers = event['headers']
    queryStringParameters = event['queryStringParameters']
    pathParameters = event['pathParameters']
    stageVariables = event['stageVariables']

    # Parse the input for the parameter values
    tmp = event['methodArn'].split(':')
    apiGatewayArnTmp = tmp[5].split('/')
    awsAccountId = tmp[4]
    region = tmp[3]
    restApiId = apiGatewayArnTmp[0]
    stage = apiGatewayArnTmp[1]
    method = apiGatewayArnTmp[2]
    resource = '/'

    if (apiGatewayArnTmp[3]):
        resource += apiGatewayArnTmp[3]

    # Perform authorization to return the Allow policy for correct parameters
    # and the 'Unauthorized' error, otherwise.

    authResponse = {}
    condition = {}
    condition['IpAddress'] = {}

    if (headers['headerauth1'] == "headerValue1" and queryStringParameters["QueryString1"]
            == "queryValue1" and stageVariables["StageVal1"] == "stageValue1"):
        response = generateAllow('me', event['methodArn'])
        print('authorized')
        return json.loads(response)
    else:
        print('unauthorized')
        raise Exception('Unauthorized') # Return a 401 Unauthorized response
        return 'unauthorized'

    # Help function to generate IAM policy


def generatePolicy(principalId, effect, resource):
    authResponse = {}
    authResponse['principalId'] = principalId
    if (effect and resource):
        policyDocument = {}
        policyDocument['Version'] = '2012-10-17'
        policyDocument['Statement'] = []
        statementOne = {}
        statementOne['Action'] = 'execute-api:Invoke'
        statementOne['Effect'] = effect
        statementOne['Resource'] = resource
        policyDocument['Statement'] = [statementOne]
        authResponse['policyDocument'] = policyDocument

    authResponse['context'] = {
        "stringKey": "stringval",
        "numberKey": 123,
        "booleanKey": True
    }

    authResponse_JSON = json.dumps(authResponse)

    return authResponse_JSON


def generateAllow(principalId, resource):
    return generatePolicy(principalId, 'Allow', resource)


def generateDeny(principalId, resource):
    return generatePolicy(principalId, 'Deny', resource)

File: oathkeeper/authorizer/authorizer.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/secretsmanager"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-secretsmanager-caching-go/secretcache"

	sdk "github.com/ory/oathkeeper-client-go"
)

func getCachedJwks() (map[string]interface{}, error) {
	secretName := "oathkeeper_jwks"

	var (
		secretCache, _ = secretcache.New()
	)
	result, err := secretCache.GetSecretString(secretName)
	if err != nil {
		return nil, err
	}

	var jwks map[string]interface{}
	if err := json.Unmarshal([]byte(result), &jwks); err != nil {
		return nil, err
	}

	return jwks, nil
}

func getStoredJwks() (map[string]interface{}, error) {
	secretName := "oathkeeper_jwks"
	secretRegion := os.Getenv("AWS_REGION")
	if os.Getenv("ENVIRONMENT") == "test" {
		secretName = "oathkeeper_jwks_test"
	}
	if secretRegion == "" {
		secretRegion = "us-west-2"
	}

	config, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion(secretRegion))
	if err != nil {
		log.Fatal(err)
	}
	svc := secretsmanager.NewFromConfig(config)
	input := &secretsmanager.GetSecretValueInput{
		SecretId:     aws.String(secretName),
		VersionStage: aws.String("AWSCURRENT"),
	}

	result, err := svc.GetSecretValue(context.TODO(), input)
	if err != nil {
		log.Fatal(err.Error())
	}

	var secretString string = *result.SecretString
	var jwks map[string]interface{}
	if err := json.Unmarshal([]byte(secretString), &jwks); err != nil {
		return nil, err
	}

	return jwks, nil
}

func getJwksData() ([]byte, error) {
	cachedJwks, err := getCachedJwks()
	if err != nil || cachedJwks == nil {
		fmt.Println("Failed to get cached JWKS. Getting from Secrets Manager...")
		storedJwks, err := getStoredJwks()
		if err != nil {
			return nil, err
		}
		return json.Marshal(storedJwks)
	}

	jwksData, err := json.Marshal(cachedJwks)
	if err != nil {
		return nil, err
	}

	return jwksData, nil
}

func setJwks() {
	// Get JWKS from AWS Secrets Manager
	jwksData, err := getJwksData()
	if err != nil {
		fmt.Printf("Failed to get JWKS data: %v\n", err)
		return
	}

	// Write JWKS to file
	jwksFilePath := "tmp/jwks.json"
	os.MkdirAll(filepath.Dir(jwksFilePath), 0755)
	jwksFile, err := os.Create(jwksFilePath)
	if err != nil {
		fmt.Printf("Failed to create JWKS file: %v\n", err)
		return
	}
	jwksFile.Write(jwksData)
	jwksFile.Close()
}

func getDecisionsRequest(event events.APIGatewayProxyRequest) sdk.ApiApiDecisionsRequest {
	configuration := sdk.NewConfiguration()

	// Set the headers needed by oathkeeper
	configuration.AddDefaultHeader("X-Forwarded-Method", event.HTTPMethod)
	configuration.AddDefaultHeader("X-Forwarded-Uri", event.Path)
	configuration.AddDefaultHeader("X-Forwarded-Host", event.Headers["Host"])
	configuration.AddDefaultHeader("X-Forwarded-Proto", "http")
	configuration.AddDefaultHeader("X-Forwarded-For", event.RequestContext.Identity.SourceIP)
	configuration.AddDefaultHeader("Authorization", event.Headers["Authorization"])

	configuration.Scheme = "http"
	configuration.Host = "0.0.0.0:4456"

	apiClient := sdk.NewAPIClient(configuration)
	request := apiClient.ApiApi.Decisions(context.Background())

	return request
}

func generatePolicy(principalID string, effect string, resource string) events.APIGatewayCustomAuthorizerResponse {
	authResponse := events.APIGatewayCustomAuthorizerResponse{
		PrincipalID: principalID,
		PolicyDocument: events.APIGatewayCustomAuthorizerPolicy{
			Version: "2012-10-17",
			Statement: []events.IAMPolicyStatement{
				{
					Action:   []string{"execute-api:Invoke"},
					Effect:   effect,
					Resource: []string{resource},
				},
			},
		},
	}

	return authResponse
}

func generateAllow(principalID string, resource string, oathkeeperResponse *http.Response) events.APIGatewayCustomAuthorizerResponse {
	policy := generatePolicy(principalID, "Allow", resource)
	policy.Context = map[string]interface{}{
		"authorizer": oathkeeperResponse.Header.Get("Authorization"),
		"x-user":     oathkeeperResponse.Header.Get("X-User"),
	}
	return policy
}

func generateDeny(principalID string, resource string) events.APIGatewayCustomAuthorizerResponse {
	return generatePolicy(principalID, "Deny", resource)
}

func init() {
	setJwks()

	// Spawn Oathkeeper Server
	cmd := exec.Command("oathkeeper", "--serve", "--config", "/etc/config/oathkeeper/config.yml")
	if err := cmd.Start(); err != nil {
		fmt.Printf("Error starting Oathkeeper server: %v\n", err)
		return
	}
}

func handler(ctx context.Context, event events.APIGatewayProxyRequest) (events.APIGatewayCustomAuthorizerResponse, error) {

	fmt.Printf("Event: %v\n", event)
	fmt.Printf("Context: %v\n", ctx)

	resp, err := getDecisionsRequest(event).Execute()
	if resp == nil || err != nil {
		fmt.Printf("Error executing request: %v\n", err)
		return generateDeny("user", event.RequestContext.Path), nil
	}

	return generateAllow("user", event.RequestContext.Path, resp), nil
}

func main() {
	// Start Lambda handler
	lambda.Start(handler)
}
